<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Merging the GenOMICC and UK Biobank Cohorts · GenOMICC Workflows</title><meta name="title" content="Merging the GenOMICC and UK Biobank Cohorts · GenOMICC Workflows"/><meta property="og:title" content="Merging the GenOMICC and UK Biobank Cohorts · GenOMICC Workflows"/><meta property="twitter:title" content="Merging the GenOMICC and UK Biobank Cohorts · GenOMICC Workflows"/><meta name="description" content="Documentation for GenOMICC Workflows."/><meta property="og:description" content="Documentation for GenOMICC Workflows."/><meta property="twitter:description" content="Documentation for GenOMICC Workflows."/><meta property="og:url" content="https://baillielab.github.io/genomicc-workflows/stable/ukb_merge/"/><meta property="twitter:url" content="https://baillielab.github.io/genomicc-workflows/stable/ukb_merge/"/><link rel="canonical" href="https://baillielab.github.io/genomicc-workflows/stable/ukb_merge/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">GenOMICC Workflows</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../walk_through/">Walk Through</a></li><li><span class="tocitem">Prerequisites</span><ul><li><a class="tocitem" href="../odap_prerequisites/">ODAP</a></li><li><a class="tocitem" href="../rap_prerequisites/">UKB RAP</a></li></ul></li><li><span class="tocitem">Workflows</span><ul><li><a class="tocitem" href="../combining_genomicc_datasets/">Combining GenOMICC Datasets</a></li><li><a class="tocitem" href="../genotypes_imputation/">GenOMICC Genotypes Imputation</a></li><li class="is-active"><a class="tocitem" href>Merging the GenOMICC and UK Biobank Cohorts</a><ul class="internal"><li><a class="tocitem" href="#1.-Uploading-Inputs"><span>1. Uploading Inputs</span></a></li><li><a class="tocitem" href="#2.-Extracting-Phenotypes"><span>2. Extracting Phenotypes</span></a></li><li><a class="tocitem" href="#3.-Merging-Cohorts"><span>3. Merging Cohorts</span></a></li></ul></li><li><a class="tocitem" href="../gwas/">GWAS</a></li></ul></li><li><span class="tocitem">For Developpers</span><ul><li><a class="tocitem" href="../development/">Development</a></li><li><a class="tocitem" href="../mock_data/">Mock Data</a></li><li><a class="tocitem" href="../julia_fns/">Index Of Julia Functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Workflows</a></li><li class="is-active"><a href>Merging the GenOMICC and UK Biobank Cohorts</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Merging the GenOMICC and UK Biobank Cohorts</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/baillielab/genomicc-workflows" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/baillielab/genomicc-workflows/blob/main/docs/src/ukb_merge.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Merging-the-GenOMICC-and-UK-Biobank-Cohorts"><a class="docs-heading-anchor" href="#Merging-the-GenOMICC-and-UK-Biobank-Cohorts">Merging the GenOMICC and UK Biobank Cohorts</a><a id="Merging-the-GenOMICC-and-UK-Biobank-Cohorts-1"></a><a class="docs-heading-anchor-permalink" href="#Merging-the-GenOMICC-and-UK-Biobank-Cohorts" title="Permalink"></a></h1><p>This workflow yields combined:</p><ul><li>genotypes in PLINK bed format: obtained by subsampling the TOPMed UKB imputed genotypes to keep the GenOMICC variants.</li><li>imputed genotypes in PLINK pgen format: ontained by merging UKB and GenOMICC TOPMed UKB imputed genotypes.</li><li>covariates containing the current age and sex of individuals</li><li>ancestry estimates: via the 1000 Genome Project</li></ul><p>It is assumed that you have installed the required dependencies as explained in <a href="../rap_prerequisites/#UKB-RAP">UKB RAP</a>.</p><h2 id="1.-Uploading-Inputs"><a class="docs-heading-anchor" href="#1.-Uploading-Inputs">1. Uploading Inputs</a><a id="1.-Uploading-Inputs-1"></a><a class="docs-heading-anchor-permalink" href="#1.-Uploading-Inputs" title="Permalink"></a></h2><p>This section describes the main inputs to the workflow that need to be uploaded to the RAP. I recommend deviating as little as possible from the following instructions to make sure things run smoothly afterwards. In this section we will populate the <code>assets/rap</code> folder in this repository that we will then upload to the RAP using the upload agent.</p><p>In particular we need to populate the <code>genomicc</code> and <code>kgp</code> subfolders as displayed below. The <code>hesin_critical_fields.txt</code> and <code>main_fields.txt</code> should already be present.</p><p><img src="../assets/rap_assets_level_1.png" alt="rap_assets_level_1"/></p><h3 id="GenOMICC-Data"><a class="docs-heading-anchor" href="#GenOMICC-Data">GenOMICC Data</a><a id="GenOMICC-Data-1"></a><a class="docs-heading-anchor-permalink" href="#GenOMICC-Data" title="Permalink"></a></h3><p>We will need the following GenOMICC data, at the present time, the name of the file does not matter since it seems impossible to reference a file by its path on the RAP, instead we will have to use file IDs: </p><ul><li>genotypes: Output by <a href="../combining_genomicc_datasets/#Combining-GenOMICC-Datasets">Combining GenOMICC Datasets</a>.</li><li>covariates: A covariate file provided by the ODAP team, currently this file contains the following columns: <code>IID</code>,<code>FID</code>,<code>PRIM_DIAGNOSIS_ODAP</code>,<code>COHORT</code>,<code>ISARIC_MAX_SEVERITY_SCORE</code>,<code>SEX_SELF_REPORTED</code>,<code>AGE_YEARS_AT_RECRUITMENT</code>,<code>SUPERPOPULATION</code>.</li><li>imputed genotypes: Output by <a href="../genotypes_imputation/#GenOMICC-Genotypes-Imputation">GenOMICC Genotypes Imputation</a>.</li></ul><p>I recommend to organise them as follows:</p><p><img src="../assets/rap_assets_genomicc.png" alt="rap_assets_genomicc"/></p><div class="admonition is-info" id="Imputed-Genotypes-8a79ebac236e8dfc"><header class="admonition-header">Imputed Genotypes<a class="admonition-anchor" href="#Imputed-Genotypes-8a79ebac236e8dfc" title="Permalink"></a></header><div class="admonition-body"><p>Keep the imputed genotypes organised in chromosomes. Only 3 chromosomes are presented above for readability but all 22 chromosomes should be in the folder.</p></div></div><h3 id="1000-Genome-Project"><a class="docs-heading-anchor" href="#1000-Genome-Project">1000 Genome Project</a><a id="1000-Genome-Project-1"></a><a class="docs-heading-anchor-permalink" href="#1000-Genome-Project" title="Permalink"></a></h3><p>You will also need the 1000 Genome Project genotypes in plink format, filtered to keep only variants matching the GenOMICC genotyped variants. This is an output of the <a href="../combining_genomicc_datasets/#Combining-GenOMICC-Datasets">Combining GenOMICC Datasets</a> workflow. They can be organised as follows:</p><p><img src="../assets/rap_assets_kgp.png" alt="rap_assets_kgp"/></p><h3 id="Other-Assets"><a class="docs-heading-anchor" href="#Other-Assets">Other Assets</a><a id="Other-Assets-1"></a><a class="docs-heading-anchor-permalink" href="#Other-Assets" title="Permalink"></a></h3><p>We also need the reference genome, which can be downloaded in the <code>assets/rap</code> folder with:</p><pre><code class="language-bash hljs">wget -O assets/rap/Homo_sapiens_assembly38.fasta https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta</code></pre><h3 id="Uploading-the-data"><a class="docs-heading-anchor" href="#Uploading-the-data">Uploading the data</a><a id="Uploading-the-data-1"></a><a class="docs-heading-anchor-permalink" href="#Uploading-the-data" title="Permalink"></a></h3><p>Since this is a lot of data, we need to use the <a href="https://documentation.dnanexus.com/downloads#installing-the-upload-agent">upload agent</a>. Asumming <code>ua</code> is in your path, run:</p><pre><code class="language-bash hljs">export AUTH_TOKEN=XXX
export PROJECT_ID=project-J0pkqyQJpYQ133JG1p2J1qzv
ua --project $PROJECT_ID --auth-token $AUTH_TOKEN --recursive --do-not-compress --folder /assets assets/rap/</code></pre><h2 id="2.-Extracting-Phenotypes"><a class="docs-heading-anchor" href="#2.-Extracting-Phenotypes">2. Extracting Phenotypes</a><a id="2.-Extracting-Phenotypes-1"></a><a class="docs-heading-anchor-permalink" href="#2.-Extracting-Phenotypes" title="Permalink"></a></h2><p>At this point, it seems impossible (or very difficult) to extract phenotypes from a WDL workflow because the source dataset is not a regular file (see <a href="https://community.ukbiobank.ac.uk/hc/en-gb/community/posts/16019555833117-How-do-I-call-the-table-exporter-applet-from-my-WDL">this</a> or <a href="https://community.ukbiobank.ac.uk/hc/en-gb/community/posts/16019577183901-I-would-like-to-access-phenotype-data-from-within-a-WDL-task-using-a-python-script-How-do-I-localize-the-database-file-as-an-input-to-the-task-I-m-getting-a-wrong-type-error">that</a>). We thus have to first run a native DNA Nexus workflow to extract some preliminary data from the source dataset. First we build the workflow:</p><pre><code class="language-bash hljs">dx build rap_workflows/export_covariates</code></pre><p>Then run it (replace the <code>DATASET_RECORD_ID</code> with the one corresponding to your project):</p><pre><code class="language-bash hljs">DATASET_RECORD_ID=record-J0pqJxjJZF8G55f99FF11JJ9
dx run -y \
-istage-J0vx360JpYQ0Jg1QJ5Zv0PFx.dataset_or_cohort_or_dashboard=$DATASET_RECORD_ID \
-istage-J0vx360JpYQ0Jg1QJ5Zv0PFx.field_names_file_txt=/assets/hesin_critical_fields.txt \
-istage-J0ygjB0JpYQJg4b985gqYkx6.dataset_or_cohort_or_dashboard=$DATASET_RECORD_ID \
-istage-J0ygjB0JpYQJg4b985gqYkx6.field_names_file_txt=/assets/main_fields.txt \
/export_covariates</code></pre><p>You can monitor the workflow on the RAP, once finished you should have two outputs in the <code>/export_covariates_outputs</code> folder.</p><h2 id="3.-Merging-Cohorts"><a class="docs-heading-anchor" href="#3.-Merging-Cohorts">3. Merging Cohorts</a><a id="3.-Merging-Cohorts-1"></a><a class="docs-heading-anchor-permalink" href="#3.-Merging-Cohorts" title="Permalink"></a></h2><p>First you need to compile the WDL workflow and upload it to the RAP. However, a workflow is compiled for specific set of inputs which need to be provided first. All inputs are provided in the <code>rap_workflows/ukb_merge/inputs.json</code> file. Unfortunately all file paths are project specific and must be manually set using their <code>project-id:file-id</code> format which is quite tedious and error prone (file-ids can be found on the RAP). The inputs are as follows:</p><ul><li><code>high_ld_regions</code>: The <code>assets/rap/exclude_b38.txt</code> file.</li><li><code>reference_genome</code>: The <code>assets/rap/Homo_sapiens_assembly38.fasta</code> file.</li><li><code>ukb_bgen_filesets</code>: The files are the imputed genotypes found in the RAP&#39;s <code>Bulk/Imputation/Imputation_from_genotype_TOPmed</code> folder.</li><li><code>ukb_covariates</code>: The output of the previous step, <code>covariates_table.csv</code>.</li><li><code>hesin_critical_table</code>: The output of the previous step, <code>critical_table.csv</code>.</li><li><code>genomicc_genotypes</code>: The typed variants in plink format in <code>assets/rap/genomicc/genotypes</code>.</li><li><code>genomicc_pgen_filesets</code>: The imputed variants in plink2 format in <code>assets/rap/genomicc/imputed</code>.</li><li><code>genomicc_covariates</code>: The covariates file in <code>assets/rap/genomicc/covariates</code>.</li><li><code>kgp_genotypes</code>: The 1000 Genome Project variants in plink format in <code>assets/rap/kgp</code>.</li></ul><p>Then compiling the workflow can be done with the following:</p><pre><code class="language-bash hljs">java -jar $DX_COMPILER_PATH compile rap_workflows/ukb_merge/workflow.wdl \
-f -project $PROJECT_ID \
-reorg \
-folder /workflows/ukb_merge \
-inputs rap_workflows/ukb_merge/inputs.json</code></pre><p>where the <code>DX_COMPILER_PATH</code> and <code>PROJECT_ID</code> have to be set appropriately. The compiler might output some warnings like <code>missing input for non-optional parameter</code> but you can ignore these.</p><div class="admonition is-warning" id="Updating-workflows-cd61d265df5bb1c0"><header class="admonition-header">Updating workflows<a class="admonition-anchor" href="#Updating-workflows-cd61d265df5bb1c0" title="Permalink"></a></header><div class="admonition-body"><p>At this point in time it seems like compiling multiple times the same workflow does not replace the old files. You will need to manually erase them from the RAP.</p></div></div><p>Then, you can run the workflow with the following command</p><pre><code class="language-bash hljs">dx run -y \
-f rap_workflows/ukb_merge/inputs.dx.json \
--priority high \
--destination /ukb_merge_outputs/ \
/workflows/ukb_merge/merge_ukb_and_genomicc</code></pre><p>All outputs will be located in the <code>ukb_merge_outputs</code> folder on the RAP.</p><ul><li><code>genomicc_ukb.merged.imputed.chr_{1..22}.{pgen,psam,pva}</code>: Merged imputed genotypes.</li><li><code>ukb_genomicc.merged.{bed,bim,fam}</code>: Merged genotypes.</li><li><code>ukb_genomicc.covariates.csv</code>: Merged covariates</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../genotypes_imputation/">« GenOMICC Genotypes Imputation</a><a class="docs-footer-nextpage" href="../gwas/">GWAS »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Tuesday 14 October 2025 15:19">Tuesday 14 October 2025</span>. Using Julia version 1.11.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
