<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Mock Data · GenOMICC Workflows</title><meta name="title" content="Mock Data · GenOMICC Workflows"/><meta property="og:title" content="Mock Data · GenOMICC Workflows"/><meta property="twitter:title" content="Mock Data · GenOMICC Workflows"/><meta name="description" content="Documentation for GenOMICC Workflows."/><meta property="og:description" content="Documentation for GenOMICC Workflows."/><meta property="twitter:description" content="Documentation for GenOMICC Workflows."/><meta property="og:url" content="https://baillielab.github.io/genomicc-workflows/stable/mock_data/"/><meta property="twitter:url" content="https://baillielab.github.io/genomicc-workflows/stable/mock_data/"/><link rel="canonical" href="https://baillielab.github.io/genomicc-workflows/stable/mock_data/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">GenOMICC Workflows</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../walk_through/">Walk Through</a></li><li><span class="tocitem">Prerequisites</span><ul><li><a class="tocitem" href="../odap_prerequisites/">ODAP</a></li><li><a class="tocitem" href="../rap_prerequisites/">UKB RAP</a></li></ul></li><li><span class="tocitem">Workflows</span><ul><li><a class="tocitem" href="../combining_genomicc_datasets/">Combining GenOMICC Datasets</a></li><li><a class="tocitem" href="../genotypes_imputation/">GenOMICC Genotypes Imputation</a></li><li><a class="tocitem" href="../ukb_merge/">Merging the GenOMICC and UK Biobank Cohorts</a></li><li><a class="tocitem" href="../gwas/">GWAS</a></li></ul></li><li><span class="tocitem">For Developpers</span><ul><li><a class="tocitem" href="../development/">Development</a></li><li class="is-active"><a class="tocitem" href>Mock Data</a><ul class="internal"><li><a class="tocitem" href="#Genotyping-Arrays"><span>Genotyping Arrays</span></a></li><li><a class="tocitem" href="#Whole-Genome-Sequencing"><span>Whole Genome Sequencing</span></a></li><li><a class="tocitem" href="#Covariates"><span>Covariates</span></a></li><li><a class="tocitem" href="#How-to-Mock"><span>How to Mock</span></a></li><li><a class="tocitem" href="#Thousands-Genomes"><span>Thousands Genomes</span></a></li></ul></li><li><a class="tocitem" href="../julia_fns/">Index Of Julia Functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">For Developpers</a></li><li class="is-active"><a href>Mock Data</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Mock Data</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/baillielab/genomicc-workflows" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/baillielab/genomicc-workflows/blob/main/docs/src/mock_data.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Mock-Data"><a class="docs-heading-anchor" href="#Mock-Data">Mock Data</a><a id="Mock-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Mock-Data" title="Permalink"></a></h1><p>In order to ease testing and development, we generate mock data that closely ressembles the original data while preserving the privacy of individuals in the cohorts. This page explains how this is done.</p><p>First but foremost, for all individuals, a mock sample ID is created. Since there is no sensitive information about individuals, these cannot be identified as long as the genetic data is perturbed. We explain below how this is done.</p><h2 id="Genotyping-Arrays"><a class="docs-heading-anchor" href="#Genotyping-Arrays">Genotyping Arrays</a><a id="Genotyping-Arrays-1"></a><a class="docs-heading-anchor-permalink" href="#Genotyping-Arrays" title="Permalink"></a></h2><ol><li>Only a subset of genetic variations are kept (e.g. 100 out of 500 000)</li><li>For each individual, each variant is resampled independently from the cohort&#39;s empirical distribution. The probability of this operation to have no effect on an individual is difficult to estimate since it depends on each variant&#39;s alleles frequencies. If all variants were different it would be (1/n_samples)^nvariants which for the lower values of <span>$nsamples=1000$</span> and <span>$nvariants=110$</span>, this is lower than <span>$10^-300$</span>.</li></ol><h2 id="Whole-Genome-Sequencing"><a class="docs-heading-anchor" href="#Whole-Genome-Sequencing">Whole Genome Sequencing</a><a id="Whole-Genome-Sequencing-1"></a><a class="docs-heading-anchor-permalink" href="#Whole-Genome-Sequencing" title="Permalink"></a></h2><p>The GVCF mock data arising from whole genome sequencing is built from a very small intersection of variants (e.g., 100) common to all genotyping arrays. Individuals are thus unidentifiable.</p><h2 id="Covariates"><a class="docs-heading-anchor" href="#Covariates">Covariates</a><a id="Covariates-1"></a><a class="docs-heading-anchor-permalink" href="#Covariates" title="Permalink"></a></h2><p>Since the covariates are not sensitive, the newly created odap identifier is simply forwarded to covariates.</p><h2 id="How-to-Mock"><a class="docs-heading-anchor" href="#How-to-Mock">How to Mock</a><a id="How-to-Mock-1"></a><a class="docs-heading-anchor-permalink" href="#How-to-Mock" title="Permalink"></a></h2><p>To run, on ODAP, assuming:</p><ul><li>The data output by Dominique is in <code>/odp-beefgs/a015/linked_data/preqc/array-pre-imputation/</code> and mounted in the singularity container in <code>/mnt/data</code></li><li>The repo is mounted in <code>/mnt/genomicc-workflows</code> (This is not necessary anymore once the code is in the container, just need to point to <code>/opt/genomicc-workflows</code>)</li></ul><pre><code class="language-bash hljs">singularity shell --bind /odp-beefgs/a015/linked_data/preqc/array-pre-imputation/:/mnt/data PATH_TO_SINGULARITY_IMAGE</code></pre><p>Then run </p><pre><code class="language-bash hljs">JULIA_DEPOT_PATH=$JULIA_DEPOT_PATH:/root/.julia julia --project=/opt/genomicc-workflows /opt/genomicc-workflows/bin/genomicc.jl</code></pre><p>I also manually:</p><ul><li>Added duplicate sample IDs to reproduce what is in the data.</li><li>Changed the position of <code>GSA-rs114361133</code> in <code>test/assets/genomicc/genotyping_arrays/mock.release_2021_2023.map</code> to be unliftable.</li></ul><h2 id="Thousands-Genomes"><a class="docs-heading-anchor" href="#Thousands-Genomes">Thousands Genomes</a><a id="Thousands-Genomes-1"></a><a class="docs-heading-anchor-permalink" href="#Thousands-Genomes" title="Permalink"></a></h2><p>The data was downloaded from the 1000GP <a href="https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/release/20130502/">FTP</a> and pruned using the <code>bin/make_thousand_genomes_filter_files.jl</code> script.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../development/">« Development</a><a class="docs-footer-nextpage" href="../julia_fns/">Index Of Julia Functions »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Tuesday 14 October 2025 15:20">Tuesday 14 October 2025</span>. Using Julia version 1.11.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
