<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Combining Datasets · SequentialGWAS.jl</title><meta name="title" content="Combining Datasets · SequentialGWAS.jl"/><meta property="og:title" content="Combining Datasets · SequentialGWAS.jl"/><meta property="twitter:title" content="Combining Datasets · SequentialGWAS.jl"/><meta name="description" content="Documentation for SequentialGWAS.jl."/><meta property="og:description" content="Documentation for SequentialGWAS.jl."/><meta property="twitter:description" content="Documentation for SequentialGWAS.jl."/><meta property="og:url" content="https://baillielab.github.io/SequentialGWAS.jl/combining_datasets/"/><meta property="twitter:url" content="https://baillielab.github.io/SequentialGWAS.jl/combining_datasets/"/><link rel="canonical" href="https://baillielab.github.io/SequentialGWAS.jl/combining_datasets/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">SequentialGWAS.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><span class="tocitem">Prereuisistes</span><ul><li><a class="tocitem" href="../odap_prerequisites/">ODAP</a></li><li><a class="tocitem" href="../rap_prerequisites/">UKB RAP</a></li></ul></li><li><span class="tocitem">Workflows</span><ul><li class="is-active"><a class="tocitem" href>Combining Datasets</a><ul class="internal"><li><a class="tocitem" href="#Inputs"><span>Inputs</span></a></li><li><a class="tocitem" href="#Running-The-Workflow"><span>Running The Workflow</span></a></li><li><a class="tocitem" href="#Outputs"><span>Outputs</span></a></li><li><a class="tocitem" href="#Workflow-Parameters"><span>Workflow Parameters</span></a></li><li><a class="tocitem" href="#Current-Limitations"><span>Current Limitations</span></a></li><li><a class="tocitem" href="#Workflow-DAG"><span>Workflow DAG</span></a></li></ul></li><li><a class="tocitem" href="../genotypes_imputation/">Genotypes Imputation</a></li><li><a class="tocitem" href="../ukb_merge/">Merging the GenOMICC and UK Biobank Cohorts</a></li><li><a class="tocitem" href="../gwas/">GWAS</a></li></ul></li><li><span class="tocitem">For Developpers</span><ul><li><a class="tocitem" href="../development/">Development</a></li><li><a class="tocitem" href="../mock_data/">Mock Data</a></li><li><a class="tocitem" href="../julia_fns/">Index Of Julia Functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Workflows</a></li><li class="is-active"><a href>Combining Datasets</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Combining Datasets</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/baillielab/sequential-gwas/blob/main/docs/src/combining_datasets.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Combining-Datasets"><a class="docs-heading-anchor" href="#Combining-Datasets">Combining Datasets</a><a id="Combining-Datasets-1"></a><a class="docs-heading-anchor-permalink" href="#Combining-Datasets" title="Permalink"></a></h1><p>This workflow combines the various genetic data sources available into a unified dataset.</p><h2 id="Inputs"><a class="docs-heading-anchor" href="#Inputs">Inputs</a><a id="Inputs-1"></a><a class="docs-heading-anchor-permalink" href="#Inputs" title="Permalink"></a></h2><p>Since GenOMICC is an ongoing project where the data is continuously collected, it came and will continue to arrive in different formats. This section describes the input data, the corresponding workflow parameters are described below.</p><h3 id="Genetic-Data"><a class="docs-heading-anchor" href="#Genetic-Data">Genetic Data</a><a id="Genetic-Data-1"></a><a class="docs-heading-anchor-permalink" href="#Genetic-Data" title="Permalink"></a></h3><p>The pipeline requires genotyping arrays and optionally whole-genome sequencing (wgs) data.</p><h4 id="Genotyping-Arrays"><a class="docs-heading-anchor" href="#Genotyping-Arrays">Genotyping Arrays</a><a id="Genotyping-Arrays-1"></a><a class="docs-heading-anchor-permalink" href="#Genotyping-Arrays" title="Permalink"></a></h4><p>There are three filesets and three corresponding subfolders:</p><ul><li>The r8 release: Corresponds to genotyping data generated before 2021. The genotyping chip was the Illumina GSA-MD-24v3-0<em>A1 and the associated genome build GRCh37. The corresponding subfolder is usually named `wp5-gwas-r8-under90excl</em>2021Sep16`.</li><li>The 2021-2023 release:  Corresponds to genotyping data generated between 2021 and 2023. It was also genotyped using the Illumina GSA-MD-24v3-0<em>A1 chip and the genome build is also GRCh37. The corresponding subfolder is usually named `20210920</em>20231206<em>QC</em>VFinal`.</li><li>The 2024-now release: Corresponds to the latest fileset. The Illumina GSA-MD-48v4-0<em>A1 chip was used and the genome build is GRCh38. The corresponding subfolder is usually named `20240604</em>20240610<em>QC</em>VFinal`.</li></ul><p>In each subfolder, there is a plink subfolder which contains the actual genotypes, the following tables summarises the above</p><table><tr><th style="text-align: center">Brief Description</th><th style="text-align: center">Period Begin</th><th style="text-align: center">Period End</th><th style="text-align: center">Genotyping Array</th><th style="text-align: center">Genome Build</th><th style="text-align: center">Directory</th><th style="text-align: center">Genotypes Prefix</th></tr><tr><td style="text-align: center">Prehistoric r8 release</td><td style="text-align: center">04/05/2020</td><td style="text-align: center">30/08/2021</td><td style="text-align: center">GSA-MD-24v3-0_A1</td><td style="text-align: center">GRC37</td><td style="text-align: center">wp5-gwas-r8-under90excl_2021Sep16</td><td style="text-align: center">PLINK<em>190921</em>0906/wp5-gwas-r8-under90excl_2021Sep16</td></tr><tr><td style="text-align: center">Before 2024 microarray</td><td style="text-align: center">20/09/2021</td><td style="text-align: center">06/12/2023</td><td style="text-align: center">GSA-MD-24v3-0_A1</td><td style="text-align: center">GRC37</td><td style="text-align: center">20210920<em>20231206</em>QC_VFinal</td><td style="text-align: center">PLINK<em>040724</em>0954/20210920<em>20231206</em>QC_VFinal</td></tr><tr><td style="text-align: center">Since 2024 microarray</td><td style="text-align: center">04/06/2024</td><td style="text-align: center">10/06/2024</td><td style="text-align: center">GSA-MD-48v4-0_A1</td><td style="text-align: center">GRC38</td><td style="text-align: center">20240604<em>20240610</em>QC_VFinal</td><td style="text-align: center">PLINK<em>040724</em>0114/20240604<em>20240610</em>QC_VFinal</td></tr></table><p>So, in the example above, the <code>R8_GENOTYPES</code> workflow parameter (see below) should point to <code>wp5-gwas-r8-under90excl_2021Sep16/PLINK_190921_0906/wp5-gwas-r8-under90excl_2021Sep16</code></p><h4 id="Whole-Genome-Sequencing"><a class="docs-heading-anchor" href="#Whole-Genome-Sequencing">Whole Genome Sequencing</a><a id="Whole-Genome-Sequencing-1"></a><a class="docs-heading-anchor-permalink" href="#Whole-Genome-Sequencing" title="Permalink"></a></h4><p>The wgs GVCF files are all located in a <code>wgs-reheadered</code> folder.</p><h3 id="External-Resources"><a class="docs-heading-anchor" href="#External-Resources">External Resources</a><a id="External-Resources-1"></a><a class="docs-heading-anchor-permalink" href="#External-Resources" title="Permalink"></a></h3><p>As well as our in-house data, the pipeline depends on external reference data. In principle these files should already be present on ODAP and there is nothing you need to do.</p><h4 id="The-1000-GP"><a class="docs-heading-anchor" href="#The-1000-GP">The 1000 GP</a><a id="The-1000-GP-1"></a><a class="docs-heading-anchor-permalink" href="#The-1000-GP" title="Permalink"></a></h4><ul><li>All VCF files and indexes present in <a href="https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working/20201028_3202_phased/">this FTP folder</a></li><li>The associated 1000 GP <a href="https://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000G_2504_high_coverage/working/1kGP.3202_samples.pedigree_info.txt">pedigree file</a>.</li></ul><p>These files should be stored in the same folder which is defined by the <code>KGP_DIR (default: /mnt/odap-beegfs/software/gwas-resources/1000-genomes-HC)</code> Nextflow parameter.</p><h4 id="GATK"><a class="docs-heading-anchor" href="#GATK">GATK</a><a id="GATK-1"></a><a class="docs-heading-anchor-permalink" href="#GATK" title="Permalink"></a></h4><ul><li>The <a href="https://storage.googleapis.com/genomics-public-data/resources/broad/hg38/v0/Homo_sapiens_assembly38.fasta">reference genome</a> published by the Broad Institute.</li></ul><p>This should be in a folder defined by the <code>GATK_DIR (default: /mnt/odap-beegfs/software/gwas-resources/gatk)</code> Nextflow parameter.</p><h2 id="Running-The-Workflow"><a class="docs-heading-anchor" href="#Running-The-Workflow">Running The Workflow</a><a id="Running-The-Workflow-1"></a><a class="docs-heading-anchor-permalink" href="#Running-The-Workflow" title="Permalink"></a></h2><p>If the previous steps have been completed successfully you can run:</p><pre><code class="language-bash hljs">taskset -c 998 nextflow run main.nf -entry CombineGeneticDatasets -c run.config -profile odap -resume -with-report -with-trace</code></pre><h2 id="Outputs"><a class="docs-heading-anchor" href="#Outputs">Outputs</a><a id="Outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Outputs" title="Permalink"></a></h2><p>All outputs are produced in <code>PUBLISH_DIR</code> (defaults to <code>results</code>), the main outputs of the workflow are:</p><ul><li><code>report.md</code>: A report of the pipeline execution</li><li><code>genotypes.aggregated.qced.final.{bed,bim,fam}</code>: The aggregated genotypes</li><li><code>covariates.inferred.csv</code>: THe covariates inferred from the genotypes (ancestry, PCs).</li></ul><h2 id="Workflow-Parameters"><a class="docs-heading-anchor" href="#Workflow-Parameters">Workflow Parameters</a><a id="Workflow-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Workflow-Parameters" title="Permalink"></a></h2><p>This is the list of all the pipeline&#39;s parameters, they can be set in the <code>run.config</code> file under the <code>params</code> section.</p><h3 id="Input-Files"><a class="docs-heading-anchor" href="#Input-Files">Input Files</a><a id="Input-Files-1"></a><a class="docs-heading-anchor-permalink" href="#Input-Files" title="Permalink"></a></h3><p>These are project specific and need to be provided:</p><ul><li><code>R8_GENOTYPES</code>: Prefix to release r8 genotypes (before 2021).</li><li><code>BEFORE_2024_GENOTYPES</code>: Prefix to genotypes released between 2021-2023.</li><li><code>SINCE_2024_GENOTYPES</code>: Prefix to genotypes released after 2024.</li><li><code>WGS_GVCFS</code> (optional): Prefix to whole genome sequencing files.</li></ul><h3 id="External-Inputs-Parameters"><a class="docs-heading-anchor" href="#External-Inputs-Parameters">External Inputs Parameters</a><a id="External-Inputs-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#External-Inputs-Parameters" title="Permalink"></a></h3><p>These are already set if you are using the <code>odap</code> profile.</p><ul><li><code>RESOURCES_DIR</code> (default: ./assets/resources&quot;): Path to all external resources.</li><li><code>KGP_DIR</code>: Path to the 1000 Genome Project specific resources (see <a href="#The-1000-GP">The 1000 GP</a>).</li><li><code>GATK_DIR</code>: Path to GATK specific resources (see <a href="#GATK">GATK</a>).</li><li><code>GRC37_TO_GRC38_CHAIN_FILE</code> (default: &quot;./assets/hg19ToHg38.over.chain.gz&quot;): Path to chain file used to liftover the GRCh37 genotypes to GRCh38.</li></ul><h3 id="QC-Parameters"><a class="docs-heading-anchor" href="#QC-Parameters">QC Parameters</a><a id="QC-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#QC-Parameters" title="Permalink"></a></h3><ul><li><code>QC_GENOTYPE_MISSING_RATE</code> (default: 0.02): Maximum missing rate per variant across all individuals. Variants above the threshold are dropped.</li><li><code>QC_INDIVIDUAL_MISSING_RATE</code> (default: 0.02): Maximum missing rate per individual across genotypes. Individuals above the threshold are dropped.</li><li><code>QC_HWE_P</code> (default: 1e-10): Used to identify potential technical artifacts and drop variants.</li><li><code>QC_HWE_K</code> (default: 0.001): Used together with <code>QC_HWE_P</code></li><li><code>PCA_APPROX</code> (default: true): Whether PCA is performed via approximation <a href="https://www.cog-genomics.org/plink/2.0/strat">see</a></li><li><code>FILTER_HIGH_LOADINGS_VARIANTS</code> (default: false): Whether to drop variants with high PCA loadings. If the loadings plots exhibits a high peak you may want to turn that on.</li><li><code>ANCESTRY_THRESHOLD</code> (default: 0.8): For each individual, the most likely ancestry estimate should be greater than this threshold otherwise the individual is marked as admixed.</li></ul><h3 id="Output-Directories-Parameters"><a class="docs-heading-anchor" href="#Output-Directories-Parameters">Output Directories Parameters</a><a id="Output-Directories-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Output-Directories-Parameters" title="Permalink"></a></h3><ul><li><code>PUBLISH_DIR</code> (default: &quot;results&quot;): Top level directory where outputs will be output.</li><li><code>KGP_PUBLISH_DIR</code> (default: &quot;results/kgp&quot;): Where 1000 Genome Project data will be output.</li><li><code>ARRAY_GENOTYPES_PUBLISH_DIR</code> (default: &quot;results/array_genotypes&quot;): Where data associated with the genotyping arrays will be output.</li><li><code>WGS_PUBLISH_DIR</code> (default: &quot;results/wgs&quot;): Where data associated with the whole-genome sequencing data will be output.</li><li><code>GATK_PUBLISH_DIR</code> (default: &quot;results/gatk&quot;): Where data associated with GATK requirements will be output.</li><li><code>MERGED_PUBLISH_DIR</code> (default: &quot;results/merged&quot;): Where the merged genetic data will be output.</li></ul><h2 id="Current-Limitations"><a class="docs-heading-anchor" href="#Current-Limitations">Current Limitations</a><a id="Current-Limitations-1"></a><a class="docs-heading-anchor-permalink" href="#Current-Limitations" title="Permalink"></a></h2><p>These are current limitations of the aggregation workflow:</p><ul><li>Only chromosomes 1 to 22 are processed.</li><li>Only bi-allelic SNPs are used.</li></ul><h2 id="Workflow-DAG"><a class="docs-heading-anchor" href="#Workflow-DAG">Workflow DAG</a><a id="Workflow-DAG-1"></a><a class="docs-heading-anchor-permalink" href="#Workflow-DAG" title="Permalink"></a></h2><iframe src="../assets/combining_datasets_dag.html" width="100%" height="2000px"></iframe></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../rap_prerequisites/">« UKB RAP</a><a class="docs-footer-nextpage" href="../genotypes_imputation/">Genotypes Imputation »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.13.0 on <span class="colophon-date" title="Monday 30 June 2025 14:42">Monday 30 June 2025</span>. Using Julia version 1.11.5.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
