<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Index Of Julia Functions · GenOMICC Workflows</title><meta name="title" content="Index Of Julia Functions · GenOMICC Workflows"/><meta property="og:title" content="Index Of Julia Functions · GenOMICC Workflows"/><meta property="twitter:title" content="Index Of Julia Functions · GenOMICC Workflows"/><meta name="description" content="Documentation for GenOMICC Workflows."/><meta property="og:description" content="Documentation for GenOMICC Workflows."/><meta property="twitter:description" content="Documentation for GenOMICC Workflows."/><meta property="og:url" content="https://baillielab.github.io/genomicc-workflows/stable/julia_fns/"/><meta property="twitter:url" content="https://baillielab.github.io/genomicc-workflows/stable/julia_fns/"/><link rel="canonical" href="https://baillielab.github.io/genomicc-workflows/stable/julia_fns/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">GenOMICC Workflows</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../walk_through/">Walk Through</a></li><li><span class="tocitem">Prerequisites</span><ul><li><a class="tocitem" href="../odap_prerequisites/">ODAP</a></li><li><a class="tocitem" href="../rap_prerequisites/">UKB RAP</a></li></ul></li><li><span class="tocitem">Workflows</span><ul><li><a class="tocitem" href="../combining_genomicc_datasets/">Combining GenOMICC Datasets</a></li><li><a class="tocitem" href="../genotypes_imputation/">GenOMICC Genotypes Imputation</a></li><li><a class="tocitem" href="../ukb_merge/">Merging the GenOMICC and UK Biobank Cohorts</a></li><li><a class="tocitem" href="../gwas/">GWAS</a></li></ul></li><li><span class="tocitem">For Developpers</span><ul><li><a class="tocitem" href="../development/">Development</a></li><li><a class="tocitem" href="../mock_data/">Mock Data</a></li><li class="is-active"><a class="tocitem" href>Index Of Julia Functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">For Developpers</a></li><li class="is-active"><a href>Index Of Julia Functions</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Index Of Julia Functions</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/baillielab/genomicc-workflows" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/baillielab/genomicc-workflows/blob/main/docs/src/julia_fns.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Index-of-Julia-Functions"><a class="docs-heading-anchor" href="#Index-of-Julia-Functions">Index of Julia Functions</a><a id="Index-of-Julia-Functions-1"></a><a class="docs-heading-anchor-permalink" href="#Index-of-Julia-Functions" title="Permalink"></a></h1><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.alive_at_assessment-Tuple{Any, Any}" href="#GenomiccWorkflows.alive_at_assessment-Tuple{Any, Any}"><code>GenomiccWorkflows.alive_at_assessment</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">alive_at_assessment(alive_60_genomicc, alive_28_isaric)</code></pre><p>The provided column names are not ideal:     - ALIVE<em>AT</em>28<em>DAYS: only concerns individuals with COHORT = ISARIC4C     - ALIVE</em>AT<em>60</em>DAYS: only concerns individuals with COHORT = GenOMICC</p><p>The retention time is not the same for the two cohorts.  The created column ALIVE<em>AT</em>ASSESSMENT informs wether an individual had died at the end of their respective assessment time (28 or 60 days).</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/merge_ukb_genomicc.jl#L147-L156">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.genotype_gvcf-NTuple{4, Any}" href="#GenomiccWorkflows.genotype_gvcf-NTuple{4, Any}"><code>GenomiccWorkflows.genotype_gvcf</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">genotype_gvcf(gvcf_file, 
    shared_variants_plink, 
    shared_variants_gatk, 
    reference_genome; 
    output_prefix=&quot;output&quot;
)</code></pre><p>Genotype a GVCF file using GATK and PLINK as follow. </p><ul><li>First attempts to genotype the GVCF file using GATK (this can fail for some reason in which case we just skip the individual).</li><li>Then converts the VCF to a PLINK bed format.</li><li>Updates the bim file with the mapped alleles from the shared variants (thi can also fail if some variants are not any of the known ref/alt in which case we also skip the individual).</li><li>Writes the new bim, bed and fam files.</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/gvcf_genotyping.jl#L96-L111">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.get_action-Tuple{Any, Any}" href="#GenomiccWorkflows.get_action-Tuple{Any, Any}"><code>GenomiccWorkflows.get_action</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Compares the variant to the 1000 GP information and returns an action to take together with a reason.</p><p>Potential actions are</p><ol><li>DROP</li><li>KEEP</li><li>FLIP</li></ol><p>Some particularly unexpected ACTION (REASON) are:</p><ul><li>&quot;KEEP (REVERSE-REF-ALT)&quot;</li><li>&quot;FLIP (COMPLEMENT-NOT-MATCHING-KGP)&quot;</li></ul><p>Because they mean the minor/major alleles are reversed in our dataset as compared to the reference KGP.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/qc_from_kgp.jl#L28-L42">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.is_severe_infection-Tuple{Any, Any}" href="#GenomiccWorkflows.is_severe_infection-Tuple{Any, Any}"><code>GenomiccWorkflows.is_severe_infection</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">is_severe_infection(row, infection_name)</code></pre><p>An individual is considered severe for an infection if it had the infection and is severely ill.  If an individual does not have the infection, it is filled missing not to contaminate downstream analyses.  They could be severe for another infection for example and are not appropriate controls.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/merge_ukb_genomicc.jl#L109-L115">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.kgp_unrelated_individuals-Tuple{Any}" href="#GenomiccWorkflows.kgp_unrelated_individuals-Tuple{Any}"><code>GenomiccWorkflows.kgp_unrelated_individuals</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>Only keeps the first individual within each family and writes them to <code>outfile</code>.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/relatedness.jl#L1-L4">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.read_bim-Tuple{Any}" href="#GenomiccWorkflows.read_bim-Tuple{Any}"><code>GenomiccWorkflows.read_bim</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">read_bim(file)</code></pre><p>Columns Description from: https://www.cog-genomics.org/plink/1.9/formats#bim</p><ul><li>Chromosome code (either an integer, or &#39;X&#39;/&#39;Y&#39;/&#39;XY&#39;/&#39;MT&#39;; &#39;0&#39; indicates unknown) or name</li><li>Variant identifier</li><li>Position in morgans or centimorgans (safe to use dummy value of &#39;0&#39;)</li><li>Base-pair coordinate (1-based; limited to 231-2)</li><li>Allele 1 (corresponding to clear bits in .bed; usually minor)</li><li>Allele 2 (corresponding to set bits in .bed; usually major)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/read_write.jl#L2-L12">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.read_fam-Tuple{Any}" href="#GenomiccWorkflows.read_fam-Tuple{Any}"><code>GenomiccWorkflows.read_fam</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">read_fam(file)</code></pre><p>Columns Description from: https://www.cog-genomics.org/plink/1.9/formats#fam</p><ul><li>Family ID (&#39;FID&#39;)</li><li>Within-family ID (&#39;IID&#39;; cannot be &#39;0&#39;)</li><li>Within-family ID of father (&#39;0&#39; if father isn&#39;t in dataset)</li><li>Within-family ID of mother (&#39;0&#39; if mother isn&#39;t in dataset)</li><li>Sex code (&#39;1&#39; = male, &#39;2&#39; = female, &#39;0&#39; = unknown)</li><li>Phenotype value (&#39;1&#39; = control, &#39;2&#39; = case, &#39;-9&#39;/&#39;0&#39;/non-numeric = missing data if case/control)</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/read_write.jl#L20-L30">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.read_map-Tuple{Any}" href="#GenomiccWorkflows.read_map-Tuple{Any}"><code>GenomiccWorkflows.read_map</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">read_map(file)</code></pre><p>Columns Description from: https://www.cog-genomics.org/plink/1.9/formats</p><ul><li>Chromosome code. PLINK 1.9 also permits contig names here, but most older programs do not.</li><li>Variant identifier</li><li>Position in morgans or centimorgans (optional; also safe to use dummy value of &#39;0&#39;)</li><li>Base-pair coordinate</li></ul></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/read_write.jl#L45-L54">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.report_qc_effect-Tuple{Any, Any}" href="#GenomiccWorkflows.report_qc_effect-Tuple{Any, Any}"><code>GenomiccWorkflows.report_qc_effect</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">report_qc_effect(input_prefix, output_prefix)</code></pre><p>Report the effect of the QC filtering process on SNPs and samples.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/report.jl#L16-L20">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.update_bim_with_mapped_alleles-Tuple{Any, Any}" href="#GenomiccWorkflows.update_bim_with_mapped_alleles-Tuple{Any, Any}"><code>GenomiccWorkflows.update_bim_with_mapped_alleles</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">update_bim_with_mapped_alleles(tmp_prefix, shared_variants_file)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/gvcf_genotyping.jl#L52-L55">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.update_variant_ids_with_map!-Tuple{Any, Any}" href="#GenomiccWorkflows.update_variant_ids_with_map!-Tuple{Any, Any}"><code>GenomiccWorkflows.update_variant_ids_with_map!</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">update_variant_ids_with_map!(bim, variant_ids_map)</code></pre><p>Updates the variant IDs to match the KGP if the following conditions are met:     - The position is present in the KGP dataset.     - The UKB alleles match the KGP alleles.</p><p>These variant IDs are marked for deletion and returned as a set.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/merge_ukb_genomicc.jl#L21-L29">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.write_map-Tuple{Any, Any}" href="#GenomiccWorkflows.write_map-Tuple{Any, Any}"><code>GenomiccWorkflows.write_map</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">write_map(file_prefix, array)</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/read_write.jl#L57-L59">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.write_release_samples_to_drop-Tuple{Any, Any}" href="#GenomiccWorkflows.write_release_samples_to_drop-Tuple{Any, Any}"><code>GenomiccWorkflows.write_release_samples_to_drop</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>We drop duplicate individuals according to the following priority:</p><p>WGS &gt; More Recent Array &gt; Older Array</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/qc_from_kgp.jl#L198-L202">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.OneTimeChecks.array_overlap-Tuple{Any, Any}" href="#GenomiccWorkflows.OneTimeChecks.array_overlap-Tuple{Any, Any}"><code>GenomiccWorkflows.OneTimeChecks.array_overlap</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">array_overlap(manifest_file1, manifest_file2)</code></pre><p>Takes two manifest files from Illumina genotyping arrays and returns the intersection of the SNP names.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/one_time_checks.jl#L56-L60">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.OneTimeChecks.identify_snps_to_flip-Tuple{Any}" href="#GenomiccWorkflows.OneTimeChecks.identify_snps_to_flip-Tuple{Any}"><code>GenomiccWorkflows.OneTimeChecks.identify_snps_to_flip</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">identify_snps_to_flip(manifest_file)</code></pre><p>According to <a href="https://knowledge.illumina.com/microarray/general/microarray-general-reference_material-list/000001565">this link</a>,  the RefStrand column in the manifest file corresponds to the standard designation for all eukaryotic organisms used by HapMap and 1000 Genomes Project. Variants with RefStrand equal to <code>-</code> need to be flipped to the + strand.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/one_time_checks.jl#L29-L35">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="GenomiccWorkflows.OneTimeChecks.make_snps_to_flip_list-Tuple{Any, Any}" href="#GenomiccWorkflows.OneTimeChecks.make_snps_to_flip_list-Tuple{Any, Any}"><code>GenomiccWorkflows.OneTimeChecks.make_snps_to_flip_list</code></a> — <span class="docstring-category">Method</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">make_snps_to_flip_list(output, manifest_file)</code></pre><p>Takes a manifest file from an Illumina genotyping array and writes a list of SNPs  that are on the - strand and need to be flipped to the + strand.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/baillielab/genomicc-workflows/blob/33a7b99dc7fdf262fd2aba54c12ed71ea953aa84/src/one_time_checks.jl#L41-L46">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../mock_data/">« Mock Data</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Tuesday 14 October 2025 15:20">Tuesday 14 October 2025</span>. Using Julia version 1.11.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
