<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>GenOMICC Genotypes Imputation · GenOMICC Workflows</title><meta name="title" content="GenOMICC Genotypes Imputation · GenOMICC Workflows"/><meta property="og:title" content="GenOMICC Genotypes Imputation · GenOMICC Workflows"/><meta property="twitter:title" content="GenOMICC Genotypes Imputation · GenOMICC Workflows"/><meta name="description" content="Documentation for GenOMICC Workflows."/><meta property="og:description" content="Documentation for GenOMICC Workflows."/><meta property="twitter:description" content="Documentation for GenOMICC Workflows."/><meta property="og:url" content="https://baillielab.github.io/genomicc-workflows/stable/genotypes_imputation/"/><meta property="twitter:url" content="https://baillielab.github.io/genomicc-workflows/stable/genotypes_imputation/"/><link rel="canonical" href="https://baillielab.github.io/genomicc-workflows/stable/genotypes_imputation/"/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">GenOMICC Workflows</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><a class="tocitem" href="../walk_through/">Walk Through</a></li><li><span class="tocitem">Prerequisites</span><ul><li><a class="tocitem" href="../odap_prerequisites/">ODAP</a></li><li><a class="tocitem" href="../rap_prerequisites/">UKB RAP</a></li></ul></li><li><span class="tocitem">Workflows</span><ul><li><a class="tocitem" href="../combining_genomicc_datasets/">Combining GenOMICC Datasets</a></li><li class="is-active"><a class="tocitem" href>GenOMICC Genotypes Imputation</a><ul class="internal"><li><a class="tocitem" href="#Inputs"><span>Inputs</span></a></li><li><a class="tocitem" href="#Running-The-Workflow"><span>Running The Workflow</span></a></li><li><a class="tocitem" href="#Imputation-Outputs"><span>Imputation Outputs</span></a></li><li><a class="tocitem" href="#Workflow-Parameters"><span>Workflow Parameters</span></a></li></ul></li><li><a class="tocitem" href="../ukb_merge/">Merging the GenOMICC and UK Biobank Cohorts</a></li><li><a class="tocitem" href="../gwas/">GWAS</a></li></ul></li><li><span class="tocitem">For Developpers</span><ul><li><a class="tocitem" href="../development/">Development</a></li><li><a class="tocitem" href="../mock_data/">Mock Data</a></li><li><a class="tocitem" href="../julia_fns/">Index Of Julia Functions</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Workflows</a></li><li class="is-active"><a href>GenOMICC Genotypes Imputation</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>GenOMICC Genotypes Imputation</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/baillielab/genomicc-workflows" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/baillielab/genomicc-workflows/blob/main/docs/src/genotypes_imputation.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="GenOMICC-Genotypes-Imputation"><a class="docs-heading-anchor" href="#GenOMICC-Genotypes-Imputation">GenOMICC Genotypes Imputation</a><a id="GenOMICC-Genotypes-Imputation-1"></a><a class="docs-heading-anchor-permalink" href="#GenOMICC-Genotypes-Imputation" title="Permalink"></a></h1><p>This workflow sends the genotypes for imputation to <a href="https://imputation.biodatacatalyst.nhlbi.nih.gov/#!pages/home">TOPMed</a>, downloads and aggregates the results per chromosome. It follows the principles provided in their <a href="https://topmedimpute.readthedocs.io/en/latest/">documentation</a>.</p><div class="admonition is-info" id="Platform-970317bc55d940d8"><header class="admonition-header">Platform<a class="admonition-anchor" href="#Platform-970317bc55d940d8" title="Permalink"></a></header><div class="admonition-body"><p>At the moment, it is impossible to run this workflow from ODAP because the TOPMed servers have not been white listed. Instead you can run it from <a href="https://digitalresearchservices.ed.ac.uk/resources/eddie">eddie</a>.</p></div></div><h2 id="Inputs"><a class="docs-heading-anchor" href="#Inputs">Inputs</a><a id="Inputs-1"></a><a class="docs-heading-anchor-permalink" href="#Inputs" title="Permalink"></a></h2><p>For this workflow to work, you will need:</p><ul><li>Genotypes: likely the output of the <a href="../combining_genomicc_datasets/#Combining-GenOMICC-Datasets">Combining GenOMICC Datasets</a> workflow.</li><li>A TOPMed token: see <a href="http://topmedimpute.readthedocs.io/en/latest/api/#authentication">this page</a>.</li><li>job ids (optional): Only to resume a crashed job.</li></ul><p>Please have a look at the wokflow parameters below for how to setup the run.</p><h2 id="Running-The-Workflow"><a class="docs-heading-anchor" href="#Running-The-Workflow">Running The Workflow</a><a id="Running-The-Workflow-1"></a><a class="docs-heading-anchor-permalink" href="#Running-The-Workflow" title="Permalink"></a></h2><p>If the previous steps have been completed successfully you can run:</p><pre><code class="language-bash hljs">nextflow run main.nf -entry Imputation -profile eddie -resume -with-report -with-trace -c run.config</code></pre><div class="admonition is-info" id="Crash-6670c248db84d6d9"><header class="admonition-header">Crash<a class="admonition-anchor" href="#Crash-6670c248db84d6d9" title="Permalink"></a></header><div class="admonition-body"><p>The <code>TOPMedImputation</code> process waits for TOPMed imputation jobs to finish, which might be longer than the maximum eddie job duration (48h) depending on the server&#39;s queue size. In that case, the workflow will crash but all the TOPMed jobs should have been submitted. Resuming the workflow will thus not work and it will try to resubmit new jobs. In order to bypass this behaviour you can pass an optional <code>TOPMED_JOBS_LIST</code> to proceed directly to the download stage. These job ids can be obtained from the TOPMed urls.</p></div></div><h2 id="Imputation-Outputs"><a class="docs-heading-anchor" href="#Imputation-Outputs">Imputation Outputs</a><a id="Imputation-Outputs-1"></a><a class="docs-heading-anchor-permalink" href="#Imputation-Outputs" title="Permalink"></a></h2><p>All outputs are produced in <code>PUBLISH_DIR</code> (defaults to <code>results</code>), the main outputs of the workflow are:</p><ul><li><code>chr_P.qced.{pgen,pvar,psam}</code>: A set of imputed genotypes, one for each chromosome in PGEN format.</li></ul><h2 id="Workflow-Parameters"><a class="docs-heading-anchor" href="#Workflow-Parameters">Workflow Parameters</a><a id="Workflow-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Workflow-Parameters" title="Permalink"></a></h2><p>This is the list of all the pipeline&#39;s parameters, they can be set in the <code>run.config</code> file under the <code>params</code> section.</p><h3 id="Inputs-Parameters"><a class="docs-heading-anchor" href="#Inputs-Parameters">Inputs Parameters</a><a id="Inputs-Parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Inputs-Parameters" title="Permalink"></a></h3><ul><li><code>GENOTYPES_PREFIX</code>: Prefix to genotypes files.</li><li><code>TOPMED_TOKEN_FILE</code>: Path to the file containing your TOPMed API token.</li></ul><h3 id="Important-Options"><a class="docs-heading-anchor" href="#Important-Options">Important Options</a><a id="Important-Options-1"></a><a class="docs-heading-anchor-permalink" href="#Important-Options" title="Permalink"></a></h3><ul><li><code>TOPMED_ENCRYPTION_PASSWORD</code>: An encryption password</li><li><code>TOPMED_JOBS_LIST</code>: If the workflow crashes and you want to resume, list the job-ids in this file (one per line). Job ids can be obtained from the job url in TOPMed.</li><li><code>N_SAMPLES_PER_IMPUTATION_JOBS</code> (default: 10000): We can only send file of less than 200000 samples to TOPMed and the server only allows 3 jobs at a time. This number ideally splits your data in 3 roughly equal batches.</li><li><code>IMPUTATION_R2_FILTER</code> (default: 0.9): Only imputed variants passing the threshold are kept, set to 0 if you want to keep them all.</li></ul><h3 id="Secondary-Options"><a class="docs-heading-anchor" href="#Secondary-Options">Secondary Options</a><a id="Secondary-Options-1"></a><a class="docs-heading-anchor-permalink" href="#Secondary-Options" title="Permalink"></a></h3><ul><li><code>TOPMED_REFRESH_RATE</code> (default: 180): The frequency (in seconds) with which the workflow will monitor the imputation process to send further jobs.</li><li><code>TOPMED_MAX_PARALLEL_JOBS</code> (default: 3): The maximum number of concurrent imputation processes, this is limited to 3 at the moment by TOPMed.</li></ul></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../combining_genomicc_datasets/">« Combining GenOMICC Datasets</a><a class="docs-footer-nextpage" href="../ukb_merge/">Merging the GenOMICC and UK Biobank Cohorts »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.14.1 on <span class="colophon-date" title="Tuesday 14 October 2025 15:20">Tuesday 14 October 2025</span>. Using Julia version 1.11.7.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
